include/seqan3/alignment/cigar_conversion.hpp

```diff
@@ -0,0 +1,392 @@
+
+    if (vector_size_at_least(2) && soft_clipping_at(last_index))
+        soft_clipping_end = cigar_count_at(last_index);
+    else if (vector_size_at_least(3) && hard_clipping_at(last_index) && soft_clipping_at(second_last_index))
+        soft_clipping_end = cigar_count_at(second_last_index);
+
+    if (soft_clipping_start + query_length + soft_clipping_end != std::ranges::size(query))
+        throw std::logic_error{"The CIGAR string indicates a query/read sequence length of "
+                               + std::to_string(soft_clipping_start + query_length + soft_clipping_end)
+                               + " but the supplied query/read sequence is of size"
+                               + std::to_string(std::ranges::size(query))};
+
+    // Assign the sequence to the alignment (a tuple of 2 gap decorators)
+    // -------------------------------------------------------------------------
+    using gapped_reference_type = gap_decorator<decltype(reference | views::slice(0, 0))>;
+    using gapped_sequence_type = gap_decorator<decltype(query | views::slice(0, 0))>;
+    using alignment_type = std::tuple<gapped_reference_type, gapped_sequence_type>;
+
+    alignment_type alignment{};
+
+    assign_unaligned(get<0>(alignment),
+                     reference
+                         | views::slice(zero_based_reference_start_position,
+                                        zero_based_reference_start_position + reference_length));
+    assign_unaligned(get<1>(alignment), query | views::slice(soft_clipping_start, soft_clipping_start + query_length));
+
+    // Insert gaps into the alignment based on the cigar vector
+    // -------------------------------------------------------------------------
+    using std::get;
+    auto current_ref_pos = std::ranges::begin(get<0>(alignment));
+    auto current_read_pos = std::ranges::begin(get<1>(alignment));
+
+    for (auto [cigar_count, cigar_operation] : cigar_vector)
+    {
+        assert(('S'_cigar_operation == cigar_operation) || ('H'_cigar_operation == cigar_operation)
+               || ('M'_cigar_operation == cigar_operation) || ('='_cigar_operation == cigar_operation)
+               || ('X'_cigar_operation == cigar_operation) || ('D'_cigar_operation == cigar_operation)
+               || ('N'_cigar_operation == cigar_operation) || ('I'_cigar_operation == cigar_operation)
+               || ('P'_cigar_operation == cigar_operation));
+
+        if (('M'_cigar_operation == cigar_operation) || ('='_cigar_operation == cigar_operation)
+            || ('X'_cigar_operation == cigar_operation))
+        {
+            std::ranges::advance(current_ref_pos, cigar_count);
+            std::ranges::advance(current_read_pos, cigar_count);
+        }
+        else if (('D'_cigar_operation == cigar_operation) || ('N'_cigar_operation == cigar_operation))
+        {
+            // insert gaps into read
+
+            assert(std::distance(current_read_pos, std::ranges::end(get<1>(alignment))) >= 0);
```

Comment:
hmmm. A deletion at the end would make no sense. 
`5M` 
```
YYYYYYYYYYYYYYYYYYYYYYYYY
XXXXX
```
`5M5D`
```
YYYYYYYYYYYYYYYYYYYYYYYYY
XXXXX-----
```
Usually gaps have a penalty, so the `5M` alignment is always better. You would just not add deletions at the end.

But technically both of the above are kind of valid.
Since a "correct" CIGAR string only has to fulfil the requirement, that the sum of `M/X/=/I` is equal to the query length.

So maybe the assert is wrong?